relay/metadata_manager.go

```diff
@@ -0,0 +1,169 @@
+package relay
+
+import (
+	"context"
+	"fmt"
+	"github.com/Layr-Labs/eigenda/core/v2"
+	"github.com/Layr-Labs/eigenda/disperser/common/v2/blobstore"
+	"github.com/Layr-Labs/eigenda/relay/cache"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"sync/atomic"
+)
+
+// Metadata about a blob. The relay only needs a small subset of a blob's metadata.
+// This struct adds caching and threading on top of blobstore.BlobMetadataStore.
+type blobMetadata struct {
+	// the size of the blob in bytes
+	blobSizeBytes uint32
+	// the size of the file containing the encoded chunks
+	totalChunkSizeBytes uint32
+	// the fragment size used for uploading the encoded chunks
+	fragmentSizeBytes uint32
+}
+
+// metadataManager encapsulates logic for fetching metadata for blobs. Utilized by the relay Server.
+type metadataManager struct {
+	ctx    context.Context
+	logger logging.Logger
+
+	// metadataStore can be used to read blob metadata from dynamoDB.
+	metadataStore *blobstore.BlobMetadataStore
+
+	// metadataCache is an LRU cache of blob metadata. Blobs that do not belong to one of the relay shards
+	// assigned to this server will not be in the cache.
+	metadataCache cache.CachedAccessor[v2.BlobKey, blobMetadata]
+
+	// shardSet is the set of shards assigned to this relay. This relay will refuse to serve metadata for blobs
+	// that are not assigned to one of these shards.
+	shardSet map[v2.RelayKey]struct{}
+
+	// concurrencyLimiter is a channel that limits the number of concurrent operations.
+	concurrencyLimiter chan struct{}
+}
+
+// newMetadataManager creates a new metadataManager.
+func newMetadataManager(
+	ctx context.Context,
+	logger logging.Logger,
+	metadataStore *blobstore.BlobMetadataStore,
+	metadataCacheSize int,
+	workPoolSize int,
+	shards []v2.RelayKey) (*metadataManager, error) {
+
+	shardSet := make(map[v2.RelayKey]struct{}, len(shards))
+	for _, shard := range shards {
+		shardSet[shard] = struct{}{}
+	}
+
+	server := &metadataManager{
+		ctx:                ctx,
+		logger:             logger,
+		metadataStore:      metadataStore,
+		shardSet:           shardSet,
+		concurrencyLimiter: make(chan struct{}, workPoolSize),
+	}
+
+	metadataCache, err := cache.NewCachedAccessor[v2.BlobKey, blobMetadata](metadataCacheSize, server.fetchMetadata)
+	if err != nil {
+		return nil, fmt.Errorf("error creating metadata cache: %w", err)
+	}
+
+	server.metadataCache = metadataCache
+
+	return server, nil
+}
+
+// metadataMap is a map of blob keys to metadata.
+type metadataMap map[v2.BlobKey]*blobMetadata
+
+// GetMetadataForBlobs retrieves metadata about multiple blobs in parallel.
+func (m *metadataManager) GetMetadataForBlobs(keys []v2.BlobKey) (*metadataMap, error) {
```

Comment:
It's probably simpler to use workerpool than handwriting the concurrency limiting for this function

Reply:
Getting conflicting recommendations here. I originally had a worker pool, but changed it at the request of @ian-shim and @mooselumph. I will start a conversation on slack to discuss.